Create contributors.json

KJLarson opened the pull request:
Making sure I understand how to use Git/GitHub in a collaborative environment by submitting a pull request that adds an empty file (I've mostly just used GitHub for my own projects and practice repositories). That seems pretty safe. Am I doing this right?

Progress on #51

hyperpress commented:
Looks good, Kari.

KJLarson commented:
Thanks!

So, now do I keep doing it like this? Working on the add-contributors branch on my forked version of the repository and creating pull requests?

rviscomi commented:
Yep, your workflow is perfect: fork this repository, work on individual features in separate branches, file a PR to merge each branch with the upstream master. 🙌 

Only suggestion I can think of is to reference the issue you're addressing in the PR description. GitHub will add annotations linking them together to make it easier to find. For now just add `Progress on #57`. In the future if the issue should be closed after the PR is merged you can add `Fixes #57` and GitHub will do it automatically.

Let's leave this PR open and have you push a couple more commits to your branch to build out the feature. We'll coordinate in #57.

------

One general tip since you're not used to working collaboratively. Remember to pull from the upstream master before you file the PR, to merge in any changes others may have added. In this case everything looks great.

Check your remotes by entering `git remote -v` and you'll probably only see your fork labelled "origin". You can add the upstream repository as another remote by entering `git remote add upstream git@github.com:HTTPArchive/almanac.httparchive.org.git`. Now, whenever you want to merge changes from this repo with your local fork, you can just enter `git pull upstream master`. And you can keep your origin up to date with `git push`.

KJLarson commented:
I think I understand how to do the contributors.html file, but the python files are tripping me up! I'm probably totally off with that stuff, but I tried to work it out. I would love feedback and/or some hints (maybe something for me to research that would help me figure it out?). Thanks!

rviscomi commented on src/contributors.py:
```diff
@@ -0,0 +1,13 @@
+contributors_json = {}
+
+
+def get_contributors():
```
We need two methods:
- one that updates the contributors_json variable based on the contents of the file
- one that returns the value of contributors_json

Rename this method to make it clearer that it's the first method. We also need to create the second method:

```suggestion
def get_contributors():
  return contributors_json

def update_contributors():
```

rviscomi commented on src/contributors.py:
```diff
@@ -0,0 +1,13 @@
+
+# Alphabatize data?
+
+get_contributors()
```
```suggestion
update_contributors()
```

rviscomi commented on src/templates/en/2019/contributors.html:
```diff
@@ -3,3 +3,16 @@
+<section>
+  <h3>{{ contributor.name }}</h3>
+  <ul>
+    <li>{{ team.name }}</li>
```
We'll need another loop to render an `<li>` for each team that the contributor is a part of.
```suggestion
    {% for team in contributor.teams %}
    <li>{{ teams[team].name }}</li>
    {% endfor %}
```

rviscomi commented:
Thanks @KJLarson this is looking great! Very close to the finish line, just left a few comments and then it should be ready to go.

rviscomi commented on src/contributors.py:
```diff
@@ -0,0 +1,13 @@
+    contributors_json = json.load(contributors_file)
+# need to return data
+
+# Alphabatize data?
```
Can remove these comments.

rviscomi commented:
@KJLarson let me know if you get stuck with any of this. I'm available to walk you through it over chat.

KJLarson commented:
Is there a way to view all the suggested changes for a file together rather than broken up into all these separate bits? I'm having a hard time making sense of all this and figuring out how it goes together. I'm sure several nights of poor quality sleep doesn't help (:-/)
Thanks!

rviscomi commented:
Try this view: https://github.com/HTTPArchive/almanac.httparchive.org/pull/57/files?utf8=%E2%9C%93&diff=split

KJLarson commented:
Another question...do you have any suggestions on how specific to be with commit messages? It seems it wouldn't be too helpful (in most cases) to just say "update filename." 

@rviscomi - I really appreciate your feedback, patience, and question answering!!! Thanks!

rviscomi commented:
> do you have any suggestions on how specific to be with commit messages? It seems it wouldn't be too helpful (in most cases) to just say "update filename."

Naming things is never easy 😁. For commits I like to have a short 1-liner describing what it is that I did. For more complex commits with many changes, I also include a list of each major change within the commit. For example:

```
add support for new widget

- optimizes widget performance on mobile
- removes redundant call to drawWidget in widget.js
- adds authentication to SuperWidget users
```

GitHub will name the commit based on the first line and add the rest of the commit message to the description. For small commits don't be afraid to be brief.

KJLarson commented:
I think I'm stuck on main.py and contributors.py. 

For contirbutors.py, I can't tell if it's done or if there are more changes needed.

For main.py, I can't tell if:
- the @app.route for contributors on line 28 is complete and I can delete the @app.route I added on line 53, OR
- if they need to be combined together, OR
- if the @app.route on line 31 isn't done, but the stuff I added isn't what it needs.

rviscomi commented on src/contributors.py:
```diff
@@ -0,0 +1,16 @@
+
+
+def get_contributors():
+  return contributors_json
```
Oh, sorry we also need to declare the global property here too.

```suggestion
  global contributors_json
  return contributors_json
```

rviscomi commented on src/main.py:
```diff
@@ -50,6 +51,17 @@ def chapter(year, chapter, lang):
+  if get_format(request) == 'json':
+    return jsonify(name=contributors.name, teams=team.name)
+
+  return render_template('contributors.html', contributors=contributor_util.get_contributors())
```
Your changes on lines 54 through 64 can be deleted. What I was saying is that the router for the Contributors page is already defined earlier in this file on lines 29-34. The only change needed is to set the `contributors` parameter to the output from the util module on line 34:

```py
return render_template('contributors.html', contributors=contributor_util.get_contributors())
``` 

Which resolves this TODO on line 33, so it can be removed:

```py
# TODO: Get contributor data and pass into the template.
```

rviscomi commented:
Once these changes are finished, the only thing left to do is to test out the page by visiting http://127.0.0.1:8080/2019/contributors with the development server running. You should see the info for the 3 contributors in the config file.

KJLarson commented:
Well, shoot. I tested it and got this:
NameError: name 'contributor_util' is not defined

It looks like it's referring to line 33 in main.py.

rviscomi commented on src/main.py:
```diff
@@ -29,8 +30,7 @@ def outline(year, lang):
 def contributors(year, lang):
-    # TODO: Get contributor data and pass into the template.
-    return render_template('%s/%s/contributors.html' % (lang, year), contributors={})
+    return render_template('%s/%s/contributors.html' % (lang, year), contributors=contributor_util.get_contributors())
```
Ah you have `contributor_util` but it's imported as `contributors_util`. Sorry if I suggested the wrong spelling!

KJLarson replied:
No more error, but now it's just a blank page. I inspected it and the div with the id="main" is empty. Do we need to import the data? I don't see anything that refers to any of the other files.

KJLarson replied:
In contributors.py, should "contributors_file" be "contributors_util"?

rviscomi replied:
I just tested your PR locally and found the issue. I'll walk you through my debug process so you can see what I did and how I identified the issues.

First, I wanted to inspect what the value of the `contributors` object is as it's passed to the template. So I added this logging statement in `main.py`:

```py
def contributors(year, lang):
    contributors = contributors_util.get_contributors()
    logging.debug('Contributors: %r' % contributors)
    return render_template('%s/%s/contributors.html' % (lang, year), contributors=contributors)
```

Then after reloading the page I see this line in the terminal:

```
DEBUG:root:Contributors: {}
```

So the object is not populated with the contents of the JSON file. Looking in `contributors.py` I noticed that the invocation on line 17 is calling `get_contributors` instead of `update_contributors`.

After I fixed that I reloaded the page again and the page was still blank. The JSON was being logged as expected this time:

```
DEBUG:root:Contributors: {'teams': {'developers': {'name': 'Web Developers'}, 'brainstormers': {'name': 'Brainstormers'}, 'authors': {'name': 'Chapter Authors'}, 'reviewers': {'name': 'Peer Reviewers'}, 'analysts': {'name': 'Data Analysts'}, 'designers': {'name': 'Web Designers'}, 'translators': {'name': 'Content Translators'}, 'editors': {'name': 'Editors'}}, 'contributors': {'argyleink': {'name': 'Adam Argyle', 'github': 'https://github.com/argyleink', 'teams': ['brainstormers']}, 'kjlarson': {'name': 'Kari Larson', 'github': 'https://github.com/KJLarson', 'teams': ['developers']}, 'rviscomi': {'name': 'Rick Viscomi', 'github': 'https://github.com/rviscomi', 'teams': ['brainstormers', 'developers']}}}
```

But looking in the source HTML I see that two sections are being rendered with empty fields:

![image](https://user-images.githubusercontent.com/1120896/60289086-a08b6980-98e3-11e9-81b6-39a1fb784a27.png)

The name of the JSON object is `contributors` but it also has properties named `teams` and `contributors`. So it's rendering a section for each of these and not the individual contributors. I modified the logic in the template to this:

```html
{% for id, contributor in contributors.contributors.items() %}
<section>
  <h3>{{ contributor.name }}</h3>
  <ul>
    {% for team in contributor.teams %}
    <li>{{ contributors.teams[team].name }}</li>
    {% endfor %}
  </ul>
</section>
{% endfor %}
```

The major changes were:

- iterating over the `contributors` object within the variable named `contributors`
- calling `.items()` on the JSON object so we can access its key/value separately
- assigning the key/value pair `id, contributor` respectively
- referencing the `teams` object within the `contributors` variable

Now I'm seeing the page render as expected:

![image](https://user-images.githubusercontent.com/1120896/60289499-8e5dfb00-98e4-11e9-96b0-0bbb5cd7abc1.png)

So go ahead and make the fixes to `contributors.py` and `contributors.html`. No need to commit the logging code, that's just for local debugging.

KJLarson replied:
I'm trying to recreate your process, and when I try to call update_contributors on line 17, I get this error:
```
line 13, in update_contributors
    with open('config/contributors.json', 'r') as contributors_file:
FileNotFoundError: [Errno 2] No such file or directory: 'config/contributors.json'
```
I cannot figure out what is happening here. I double-checked the spelling and path a thousand times, and they seem right. Did you run into anything like this?

KJLarson replied:
I added src/ in front of config/contributors.json and that seemed to work. That confuses me. I thought that when we use a relative path, we can just type a name that is on a sibling level in the directory tree. Doesn't this go up and then back down? Does it work differently here?

rviscomi replied:
When we deploy the app to production, it takes everything under `src` and runs it on App Engine. So there will be no `src` directory.

When you're running it locally are you running `python src/main.py` or from the src directory are you running `python main.py`? The latter is the correct way. We should update our README.

KJLarson replied:
Aha!

rviscomi commented on src/contributors.py:
```diff
@@ -0,0 +1,17 @@
+def update_contributors():
+  global contributors_json
+  
+  with open('src/config/contributors.json', 'r') as contributors_file:
```
```suggestion
  with open('config/contributors.json', 'r') as contributors_file:
```

rviscomi commented:
![image](https://user-images.githubusercontent.com/1120896/60346410-14338200-9989-11e9-85bf-12213d68cc75.png)


Ok @KJLarson there comes a time in every open source contributor's life when they have to face the dreaded merge conflict. For you, today is that day! 😄

 #60 was just merged, which affects `main.py`. You've also made changes to that file, the same lines in fact, so now git doesn't know how to properly merge them after doing a `git pull git://github.com/HTTPArchive/almanac.httparchive.org.git master` (or similar, depending on how you set up your remote repos locally).

After syncing to master you'll see this error message in the terminal:

```
Auto-merging src/main.py
CONFLICT (content): Merge conflict in src/main.py
Automatic merge failed; fix conflicts and then commit the result.
```

If you open up `main.py` you'll see this:

```
def contributors(year, lang):
<<<<<<< HEAD
    # TODO: Get contributor data and pass into the template.
    return render_template('%s/%s/contributors.html' % (lang, year), contributors={}, supported_languages=supported_languages)
=======
    contributors=contributors_util.get_contributors()
    return render_template('%s/%s/contributors.html' % (lang, year), contributors=contributors)
>>>>>>> 4ec6c11f9ef25c73c71cff37f1d5e80a663cecd4
```

There are two sections to this, the part between HEAD and === (the latest contents of the file) and the part between === and >>> (your changes to the old file). To resolve the conflict, we need to merge the two sections while preserving old and new changes. Here's how I would merge them:

```py
def contributors(year, lang):
    contributors=contributors_util.get_contributors()
    return render_template('%s/%s/contributors.html' % (lang, year), contributors=contributors, supported_languages=supported_languages)
```

(Note that none of the <<< === >>> scaffolding should exist in the code after merging)

After saving that change, if you run `git status` you'll see this:

```
Unmerged paths:
  (use "git add <file>..." to mark resolution)

	both modified:   src/main.py
```

So run `git add src/main.py` and `git commit` which will generate a merge commit. You don't need to modify the commit message, it already contains info about what you're merging. After pushing, the conflict should be resolved and the PR will be ready to merge with the master branch.

rviscomi commented on src/main.py:
```diff
@@ -30,8 +31,9 @@ def outline(year, lang):
 def contributors(year, lang):
-    # TODO: Get contributor data and pass into the template.
-    return render_template('%s/%s/contributors.html' % (lang, year), contributors={}, supported_languages=supported_languages)
+def contributors(year, lang):
```
```suggestion
```

rviscomi commented on src/main.py:
```diff
@@ -30,8 +31,9 @@ def outline(year, lang):
 def contributors(year, lang):
-    # TODO: Get contributor data and pass into the template.
-    return render_template('%s/%s/contributors.html' % (lang, year), contributors={}, supported_languages=supported_languages)
+def contributors(year, lang):
```
Accidentally copied this line. Should be removed.

KJLarson replied:
Oops! Thanks!

rviscomi commented on src/main.py:
```diff
@@ -31,9 +30,7 @@ def outline(year, lang):
 @validate
 def contributors(year, lang):
-def contributors(year, lang):
-    contributors=contributors_util.get_contributors()
```
Whoops. This commit totally rolls back the progress on this issue. Can you add these changes back in and only delete the duplicate `contributors` function definition above?

KJLarson replied:
🤦‍♀️ Oh my, I think I'm getting confused. It's tricky looking at the changes and suggestions with the green and the red!
Is this what we want?:
```
def contributors(year, lang):
    contributors=contributors_util.get_contributors()
    return render_template('%s/%s/contributors.html' % (lang, year), contributors=contributors, supported_languages=supported_languages)
```

rviscomi replied:
Yes, that's it! Also make sure to keep the "import contributors" bit above.

KJLarson replied:
Where is the "import contributors" bit?

Thanks for your patience!

rviscomi replied:
![image](https://user-images.githubusercontent.com/1120896/60386906-be82d680-9a69-11e9-86af-e8761eb9f069.png)

This part is required because the `contributors` method calls `contributors_util.get_contributors()`.
